compiler/rustc_typeck/src/expr_use_visitor.rs

```diff
@@ -228,7 +233,66 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
+                //      - If the value of the expression will be ever read
+                //      - i.e. it contains more than just a `_` being matched on
+                fn check_for_no_wild_card(pat: &Pat<'hir>) -> bool {
+                    match pat.kind {
+                        PatKind::Wild => false,
+                        PatKind::Path(_) | PatKind::Lit(_) | PatKind::Range(_, _, _) => true,
+                        PatKind::Binding(_, _, _, opt_sub_pat) => {
+                            if let Some(sub_pat) = opt_sub_pat {
+                                check_for_no_wild_card(sub_pat.clone())
+                            } else {
+                                true
+                            }
+                        }
+                        PatKind::Struct(_, sub_pats, _) => {
+                            let mut res = false;
+                            for sub_pat in sub_pats.iter() {
+                                res = res || check_for_no_wild_card(sub_pat.pat.clone());
+                            }
+                            res
+                        }
+                        PatKind::TupleStruct(_, sub_pats, _)
+                        | PatKind::Or(sub_pats)
+                        | PatKind::Tuple(sub_pats, _) => {
+                            let mut res = false;
+                            for sub_pat in sub_pats.iter() {
+                                res = res || check_for_no_wild_card(sub_pat.clone());
+                            }
+                            res
+                        }
+                        PatKind::Box(sub_pat) | PatKind::Ref(sub_pat, _) => {
+                            check_for_no_wild_card(sub_pat)
+                        }
+                        PatKind::Slice(sub_pats1, opt_sub_pat, sub_pats3) => {
+                            let mut res = false;
+                            for sub_pat in sub_pats1.iter() {
+                                res = res || check_for_no_wild_card(sub_pat.clone());
+                            }
+                            if let Some(sub_pat) = opt_sub_pat {
+                                res = res || check_for_no_wild_card(sub_pat.clone());
+                            }
+                            for sub_pat in sub_pats3.iter() {
+                                res = res || check_for_no_wild_card(sub_pat.clone());
+                            }
+                            res
+                        }
+                    }
+                }
+                debug!("RFC2229 = expr_use_visitor.rs ExprKind::Match");
+                let mut res = false;
+                for arm in arms.iter() {
+                    res = res || check_for_no_wild_card(arm.pat.clone());
+                }
+
                 let discr_place = return_if_err!(self.mc.cat_expr(&discr));
+
+                if res == true {
+                    self.delegate.fake_read(discr_place.clone());
```

Comment:
This needs to be `self.borrow_expr(&disc, Imm ... );` remove the call on line 296. Fake read for this is already introduced in `walk_pat`